Fix size of maskCanvas to draw image masks correctly

JT-External opened the pull request:
The current code creates a Canvas for drawing an ImageMask using only the top-left and bottom-right coordinates of the ImageMask. This code cannot allocate an area on the Canvas to draw the entire ImageMask when the ImageMask is rotated. Therefore, I modified the code to use the four corner coordinates of the ImageMask when creating the Canvas. I also adjusted the offset for drawing on the Canvas according to this modification.

This is a fix for #17147 .

JT-External committed:
Fix size of maskCanvas to draw image masks correctly

Use existing helper to calculate the Box

Co-authored-by: Jonas Jenwald <jonas.jenwald@gmail.com>

Ensure that there are non-zero

Co-authored-by: Jonas Jenwald <jonas.jenwald@gmail.com>

Add a reference test for #17147

Snuffleupagus commented on src/display/canvas.js:
```diff
@@ -1313,11 +1313,20 @@ class CanvasGraphics {
-    const rect = Util.normalizeRect([cord1[0], cord1[1], cord2[0], cord2[1]]);
-    const drawnWidth = Math.round(rect[2] - rect[0]) || 1;
-    const drawnHeight = Math.round(rect[3] - rect[1]) || 1;
+    const cords = [
+      [0, 0],
+      [width, height],
+      [0, height],
+      [width, 0],
+    ].map(cord => Util.applyTransform(cord, maskToCanvas));
+    const xList = cords.map(cord => cord[0]);
+    const yList = cords.map(cord => cord[1]);
+    const minX = Math.min(...xList);
+    const minY = Math.min(...yList);
+    const maxX = Math.max(...xList);
+    const maxY = Math.max(...yList);
```
Unless I'm missing something, it looks like this code is re-implementing the following helper?
https://github.com/mozilla/pdf.js/blob/5f0e5609499374c01cb9ee534f2d44bbd2ae14fb/src/shared/util.js#L725-L738

Can't this be written as follows instead?
```suggestion
    const [minX, minY, maxX, maxY] = Util.getAxialAlignedBoundingBox(
      [0, 0, width, height],
      maskToCanvas
    );
```

Snuffleupagus commented on src/display/canvas.js:
```diff
@@ -1313,11 +1313,20 @@ class CanvasGraphics {
+    const minY = Math.min(...yList);
+    const maxX = Math.max(...xList);
+    const maxY = Math.max(...yList);
+    const drawnWidth = Math.round(maxX - minX);
+    const drawnHeight = Math.round(maxY - minY);
```
The old code ensured that these are non-zero, why was that removed?
```suggestion
    const drawnWidth = Math.round(maxX - minX) || 1;
    const drawnHeight = Math.round(maxY - minY) || 1;
```

Snuffleupagus reviewed (changes requested):
Thanks for the patch; I've left a couple of inline comments below.

Can you please also add a reference-test?
For an example of how that's done, see e.g. https://github.com/mozilla/pdf.js/commit/459d26edec951fcc049928d3a07f8c933a26d3e5#diff-bbeef90df04bff7617f7ef59e3f37cac92192c5169bc500d857bd6fb107fcabd (and below).

Snuffleupagus commented:
Please squash the commits into one (we don't use separate fixup commits in this projects); see https://github.com/mozilla/pdf.js/wiki/Squashing-Commits

JT-External commented:
Thank you for your review and feedback. I appreciate your time and effort. 
I have addressed your inline comments and made the necessary changes. I have also added a reference-test as you requested. 

> Unless I'm missing something, it looks like this code is re-implementing the following helper?

I was not aware of its existence. I have updated my code to use it instead of re-implementing it.

> The old code ensured that these are non-zero, why was that removed?

I removed it by mistake. I have restored it in the latest commit. Thank you for catching that error.

Please let me know if you have any further suggestions or questions. Thank you again for your support.

calixteman commented:
/botio test

Snuffleupagus reviewed (approved):
r=me, thank you!

Snuffleupagus commented:
/botio makeref

marco-c commented:
Thanks for the contribution!

calixteman commented:
/botio-windows makeref
